### PR TITLE
[test] add coverage to rsa sign_tests

### PR DESCRIPTION
**Pull request by meilu-git**

### CODE REVIEW

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 1]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+
+    let mut hash_algo = HsmHashAlgo::Sha256;
+    let hash =
+        HsmHasher::hash_vec(&session, &mut hash_algo, b"hello").expect("Failed to hash message");
+
+    let mut algo = HsmRsaSignAlgo::with_pkcs1_padding(hash_algo);
+
+    let result = HsmVerifier::verify(&mut algo, &pub_key, &hash, &[]);
+
+    assert!(result.is_err());
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: We intentionally treat malformed inputs (like empty signature) as invalid usage and return Err. This is consistent with the rest of the negative tests asserting is_err(). Accepting Ok(false) would weaken validation semantics.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 2]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+    let mut sign_algo = HsmRsaSignAlgo::with_pkcs1_padding(hash_algo);
+    let sig =
+        HsmSigner::sign_vec(&mut sign_algo, &priv_key, &hash).expect("Failed to sign message");
+
+    // Verify with PSS
+    let mut verify_algo = HsmRsaSignAlgo::with_pss_padding(hash_algo, 32);
+
+    let result = HsmVerifier::verify(&mut verify_algo, &pub_key, &hash, &sig);
+
+    assert!(result.is_err());
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: We intentionally treat malformed inputs (like empty signature) as invalid usage and return Err. This is consistent with the rest of the negative tests asserting is_err(). Accepting Ok(false) would weaken validation semantics.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 3]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+        HsmHasher::hash_vec(&session, &mut hash_algo, b"hello").expect("Failed to hash message");
+
+    let mut algo = HsmRsaSignAlgo::with_pkcs1_padding(hash_algo);
+    let mut sig = HsmSigner::sign_vec(&mut algo, &priv_key, &hash).expect("Failed to sign message");
+
+    sig.extend_from_slice(&[0u8; 10]); // make too large
+
+    let result = HsmVerifier::verify(&mut algo, &pub_key, &hash, &sig);
+
+    assert!(result.is_err());
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: We intentionally treat malformed inputs (like empty signature) as invalid usage and return Err. This is consistent with the rest of the negative tests asserting is_err(). Accepting Ok(false) would weaken validation semantics.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 4]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+        HsmHasher::hash_vec(&session, &mut hash_algo, b"msg1").expect("Failed to hash message");
+    let hash2 =
+        HsmHasher::hash_vec(&session, &mut hash_algo, b"msg2").expect("Failed to hash message");
+
+    let mut algo = HsmRsaSignAlgo::with_pkcs1_padding(hash_algo);
+    let sig = HsmSigner::sign_vec(&mut algo, &priv_key, &hash1).expect("Failed to sign message");
+
+    let result = HsmVerifier::verify(&mut algo, &pub_key, &hash2, &sig);
+
+    assert!(result.is_err());
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: We intentionally treat malformed inputs (like empty signature) as invalid usage and return Err. This is consistent with the rest of the negative tests asserting is_err(). Accepting Ok(false) would weaken validation semantics.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 5]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+    let mut hash_algo = HsmHashAlgo::Sha256;
+    let hash =
+        HsmHasher::hash_vec(&session, &mut hash_algo, b"hello").expect("Failed to hash message");
+
+    let mut algo = HsmRsaSignAlgo::with_pkcs1_padding(hash_algo);
+    let sig = HsmSigner::sign_vec(&mut algo, &priv1, &hash).expect("Failed to sign message");
+
+    let result = HsmVerifier::verify(&mut algo, &pub2, &hash, &sig);
+
+    assert!(result.is_err());
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: We intentionally treat malformed inputs (like empty signature) as invalid usage and return Err. This is consistent with the rest of the negative tests asserting is_err(). Accepting Ok(false) would weaken validation semantics.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 6]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: We intentionally treat malformed inputs (like empty signature) as invalid usage and return Err. This is consistent with the rest of the negative tests asserting is_err(). Accepting Ok(false) would weaken validation semantics.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 7]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+
+    let mut sign_algo = HsmRsaSignAlgo::with_pss_padding(hash_algo, 32);
+    let sig =
+        HsmSigner::sign_vec(&mut sign_algo, &priv_key, &hash).expect("Failed to sign message");
+
+    let mut verify_algo = HsmRsaSignAlgo::with_pss_padding(hash_algo, 20);
+
+    let result = HsmVerifier::verify(&mut verify_algo, &pub_key, &hash, &sig);
+
+    assert!(result.is_err());
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: We intentionally treat malformed inputs (like empty signature) as invalid usage and return Err. This is consistent with the rest of the negative tests asserting is_err(). Accepting Ok(false) would weaken validation semantics.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 8]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+/// Ensure unwrap fails when private key lacks sign capability
+#[session_test]
+fn test_rsa_sign_without_permission_fails(session: HsmSession) {
+    let priv_key = RsaPrivateKey::generate(256).expect("Failed to generate RSA Key");
+    let der = priv_key.to_vec().expect("Failed to export RSA Key");
+
+    let (unwrap_priv, unwrap_pub) = get_rsa_unwrapping_key_pair(&session);
+
+    // Missing can_sign(true)
+    let priv_props = HsmKeyPropsBuilder::default()
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs:368]: Good catch — this test is validating unwrap-time key property enforcement, not sign-time failure. renamed it to reflect that.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 9]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+
+    let mut sign_algo = HsmRsaSignAlgo::with_pss_padding(hash_algo, 32);
+    let sig =
+        HsmSigner::sign_vec(&mut sign_algo, &priv_key, &hash).expect("Failed to sign message");
+
+    let mut verify_algo = HsmRsaSignAlgo::with_pkcs1_padding(hash_algo);
+
+    let result = HsmVerifier::verify(&mut verify_algo, &pub_key, &hash, &sig);
+
+    assert!(result.is_err());
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: We intentionally treat malformed inputs (like empty signature) as invalid usage and return Err. This is consistent with the rest of the negative tests asserting is_err(). Accepting Ok(false) would weaken validation semantics.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 10]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+    let mut hash_algo = HsmHashAlgo::Sha256;
+    let hash =
+        HsmHasher::hash_vec(&session, &mut hash_algo, b"hello").expect("Failed to hash message");
+
+    let mut algo = HsmRsaSignAlgo::with_pss_padding(hash_algo, 32);
+
+    let sig1 = HsmSigner::sign_vec(&mut algo, &priv_key, &hash).expect("Failed to sign message");
+    let sig2 = HsmSigner::sign_vec(&mut algo, &priv_key, &hash).expect("Failed to sign message");
+
+    assert_ne!(sig1, sig2); // critical property of PSS
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: resolved per suggestions

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 11]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+    // valid signature
+    let valid_hash = vec![0xAA; 32];
+    let sig = HsmSigner::sign_vec(&mut algo, &priv_key, &valid_hash).unwrap();
+
+    // ❌ invalid hash
+    let empty_hash = vec![];
+
+    let result = HsmVerifier::verify(&mut algo, &pub_key, &empty_hash, &sig);
+
+    assert!(result.is_err());
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: We intentionally treat malformed inputs (like empty signature) as invalid usage and return Err. This is consistent with the rest of the negative tests asserting is_err(). Accepting Ok(false) would weaken validation semantics.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 12]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+        HsmHasher::hash_vec(&session, &mut hash_algo, b"hello").expect("Failed to hash message");
+
+    let mut algo = HsmRsaSignAlgo::with_pkcs1_padding(hash_algo);
+    let mut sig = HsmSigner::sign_vec(&mut algo, &priv_key, &hash).expect("Failed to sign message");
+
+    sig[0] ^= 0xFF; // corrupt signature
+
+    let result = HsmVerifier::verify(&mut algo, &pub_key, &hash, &sig);
+
+    assert!(result.is_err());
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: We intentionally treat malformed inputs (like empty signature) as invalid usage and return Err. This is consistent with the rest of the negative tests asserting is_err(). Accepting Ok(false) would weaken validation semantics.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 13]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+        HsmHasher::hash_vec(&session, &mut hash_algo, b"hello").expect("Failed to hash message");
+
+    let mut algo = HsmRsaSignAlgo::with_pkcs1_padding(hash_algo);
+    let mut sig = HsmSigner::sign_vec(&mut algo, &priv_key, &hash).expect("Failed to sign message");
+
+    sig.truncate(sig.len() / 2);
+
+    let result = HsmVerifier::verify(&mut algo, &pub_key, &hash, &sig);
+
+    assert!(result.is_err());
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: We intentionally treat malformed inputs (like empty signature) as invalid usage and return Err. This is consistent with the rest of the negative tests asserting is_err(). Accepting Ok(false) would weaken validation semantics.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 14]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+
+    let mut sign_algo = HsmRsaSignAlgo::with_pkcs1_padding(hash_algo1);
+    let sig =
+        HsmSigner::sign_vec(&mut sign_algo, &priv_key, &hash).expect("Failed to sign message");
+
+    let mut verify_algo = HsmRsaSignAlgo::with_pkcs1_padding(hash_algo2);
+
+    let result = HsmVerifier::verify(&mut verify_algo, &pub_key, &hash, &sig);
+
+    assert!(result.is_err());
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: We intentionally treat malformed inputs (like empty signature) as invalid usage and return Err. This is consistent with the rest of the negative tests asserting is_err(). Accepting Ok(false) would weaken validation semantics.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 15]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+
+    let valid_hash = vec![0xAA; 32];
+    let sig =
+        HsmSigner::sign_vec(&mut algo, &priv_key, &valid_hash).expect("Failed to sign message");
+
+    let bad_hash = vec![0xAA; 10]; // invalid length
+
+    let result = HsmVerifier::verify(&mut algo, &pub_key, &bad_hash, &sig);
+
+    assert!(result.is_err());
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: We intentionally treat malformed inputs (like empty signature) as invalid usage and return Err. This is consistent with the rest of the negative tests asserting is_err(). Accepting Ok(false) would weaken validation semantics.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 16]
```diff
@@ -205,3 +194,473 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+        .class(HsmKeyClass::Public)
+        .key_kind(HsmKeyKind::Rsa)
+        .bits(2048)
+        .can_verify(true)
+        .build()
+        .unwrap();
+
+    let mut wrap_algo = HsmRsaAesWrapAlgo::new(HsmHashAlgo::Sha256, 32);
+    let wrapped = HsmEncrypter::encrypt_vec(&mut wrap_algo, &unwrap_pub, &der)
+        .expect("Failed to encrypt AES key");
```
[the remark this answers is not in the document]
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs]: resolved per suggestions.

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 17]
```diff
@@ -205,3 +210,530 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+    let sig =
+        HsmSigner::sign_vec(&mut sign_algo, &priv_key, &hash).expect("Failed to sign message");
+
+    let mut verify_algo = HsmRsaSignAlgo::with_pss_padding(hash_algo, 20);
+
+    let result = HsmVerifier::verify(&mut verify_algo, &pub_key, &hash, &sig);
+
+    assert!(
+        !matches!(result, Ok(true)),
+        "Verification should not succeed"
```
Comment (rajesh-gali) [inline api/tests/src/algo/rsa/sign_tests.rs:355]: should we check for Ok(false)? Because, test shouldn't generate any Err(). and !matches macro treats Err also as valid condition

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 18]
```diff
@@ -205,3 +210,530 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+    );
+}
+
+/// Ensure unwrap fails when private key lacks sign capability
+#[session_test]
+fn test_rsa_unwrap_without_sign_permission_fails(session: HsmSession) {
```
Comment (rajesh-gali) [inline api/tests/src/algo/rsa/sign_tests.rs:361]: Please add a negative test to validate sign with a key.can_sign set to false

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 19]
```diff
@@ -205,3 +210,530 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+
+    assert!(
+        !matches!(result, Ok(true)),
+        "Verification should not succeed"
+    );
+}
+
+/// Ensure verification fails when signature is corrupted
+#[session_test]
+fn test_rsa_verify_modified_signature_fails(session: HsmSession) {
```
Comment (rajesh-gali) [inline api/tests/src/algo/rsa/sign_tests.rs:248]: can we also add a test to check all 0, all 0xff signature?

File: api/tests/src/algo/rsa/sign_tests.rs [hunk 20]
```diff
@@ -205,3 +210,530 @@ fn test_rsa_4096_pss_sign_verify(session: HsmSession) {
...
+
+    assert!(
+        !matches!(result, Ok(true)),
+        "Verification should not succeed"
+    );
+}
+
+/// Ensure PSS verification fails with different salt length
+#[session_test]
+fn test_rsa_pss_salt_len_mismatch_fails(session: HsmSession) {
```
Comment (rajesh-gali) [inline api/tests/src/algo/rsa/sign_tests.rs:333]: please add a test with salt_len > hash_len, maximum valid salt len

**Review by rajesh-gali — CHANGES_REQUESTED** [review]

File: api/tests/src/algo/rsa/sign_tests.rs (hunk 20 above)
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs:333]: added two cases .

File: api/tests/src/algo/rsa/sign_tests.rs (hunk 19 above)
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs:248]: added

File: api/tests/src/algo/rsa/sign_tests.rs (hunk 18 above)
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs:361]: I remember now  I was originally wanted to do this. but  The current implementation enforces can_sign at unwrap time — keys without can _sign cannot be created. So signing with such a key is not possible.  thus I changed it to existing tests of test_rsa_unwrap_without_sign_permission_fails

File: api/tests/src/algo/rsa/sign_tests.rs (hunk 17 above)
Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs:355]: he behavior is currently platform-dependent: Ubuntu returns err,  while Windows returns  ok( false) for verification failures.  
If we assert only ok( false) , the test will fail on Ubuntu.  Copilot suggested to check for !matches!(result, Ok(true)),  so OK( false) err is accepted , ok( Ture) will fail

Comment (meilu-git) [reply api/tests/src/algo/rsa/sign_tests.rs:355]: based on discussion, will open a JIRA for this.

**Review by rajesh-gali — APPROVED** [review]
